client/allocrunnerv2/taskrunner/stats_hook.go

```diff
@@ -0,0 +1,117 @@
+package taskrunner
+
+import (
+	"context"
+	"strings"
+	"sync"
+	"time"
+
+	hclog "github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/nomad/client/allocrunnerv2/interfaces"
+	"github.com/hashicorp/nomad/client/driver"
+	cstructs "github.com/hashicorp/nomad/client/structs"
+)
+
+type StatsUpdater interface {
+	UpdateStats(*cstructs.TaskResourceUsage)
+}
+
+type statsHook struct {
+	updater  StatsUpdater
+	interval time.Duration
+
+	// stopCh is closed by Exited
+	stopCh chan struct{}
+
+	mu sync.Mutex
+
+	logger hclog.Logger
+}
+
+func newStatsHook(su StatsUpdater, interval time.Duration, logger hclog.Logger) *statsHook {
+	h := &statsHook{
+		updater:  su,
+		interval: interval,
+	}
+	h.logger = logger.Named(h.Name())
+	return h
+}
+
+func (*statsHook) Name() string {
+	return "stats_hook"
+}
+
+func (h *statsHook) Poststart(ctx context.Context, req *interfaces.TaskPoststartRequest, _ *interfaces.TaskPoststartResponse) error {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+
+	// This shouldn't happen, but better safe than risk leaking a goroutine
+	if h.stopCh != nil {
+		h.logger.Debug("poststart called twice without exiting between")
+		close(h.stopCh)
+	}
+
+	h.stopCh = make(chan struct{})
+	go h.collectResourceUsageStats(h.logger, req.DriverStats, h.stopCh)
+
+	return nil
+}
+
+func (h *statsHook) Exited(context.Context, *interfaces.TaskExitedRequest, *interfaces.TaskExitedResponse) error {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+
+	if h.stopCh == nil {
+		// No stats running
+		return nil
+	}
+
+	// Close chan to stop stats collection
+	close(h.stopCh)
+
+	// Clear chan so we don't double close for any reason
+	h.stopCh = nil
+
+	return nil
+}
+
+// collectResourceUsageStats starts collecting resource usage stats of a Task.
+// Collection ends when the passed channel is closed
+func (h *statsHook) collectResourceUsageStats(logger hclog.Logger, handle interfaces.DriverStats, stopCh <-chan struct{}) {
+	// start collecting the stats right away and then start collecting every
+	// collection interval
+	next := time.NewTimer(0)
+	defer next.Stop()
+	for {
+		select {
+		case <-next.C:
+			// Reset the timer
+			next.Reset(h.interval)
+
+			// Collect stats from driver
+			ru, err := handle.Stats()
+			if err != nil {
+				// Check if the driver doesn't implement stats
+				if err.Error() == driver.DriverStatsNotImplemented.Error() {
+					h.logger.Debug("driver does not support stats")
```

Comment:
Trace?

Reply:
Or can we add Stats to `Driver.Abilities()` and let the hook check that before spawning this goroutine?

Reply:
All drivers currently implement Stats so this is unused. If we want to require plugins to implement stats we can drop this code.

Otherwise we probably want to add this to Driver Abilities/Capabilities.